File: src/morie/fn/sarla.py
"""Spatial autoregressive lag model (SAR lag, ML)."""
import numpy as np
from scipy import optimize
from ._richresult import RichResult

__all__ = ["spatial_ar_lag"]


def spatial_ar_lag(x, y, w):
    """
    SAR lag model:
        Y = rho W Y + X beta + eps,   eps ~ N(0, sigma2 I).

    Concentrated log-likelihood in rho (Anselin 1988; Schabenberger
    & Gotway 2005, Ch 7):

        ll(rho) = -n/2 log(2 pi sigma2_hat) + log|I - rho W| - n/2
        e0 = M y,  e1 = M W y,   M = I - X (X'X)^{-1} X'
        sigma2_hat(rho) = (e0 - rho e1)' (e0 - rho e1) / n.

    Parameters
    ----------
    x : array-like, shape (n, p) -- design matrix (intercept explicit).
    y : array-like, shape (n,)
    w : array-like, shape (n, n)

    Returns
    -------
    RichResult with payload: estimate (beta), se, rho, sigma2, n, method.
    """
    X = np.asarray(x, dtype=float)
    if X.ndim == 1:
        X = X.reshape(-1, 1)
    y = np.asarray(y, dtype=float).ravel()
    W = np.asarray(w, dtype=float)
    n, p = X.shape
    if y.size != n or W.shape != (n, n):
        raise ValueError("shape mismatch among x, y, w")
    I = np.eye(n)
    XtX_inv = np.linalg.inv(X.T @ X)
    M = I - X @ XtX_inv @ X.T
    e0 = M @ y
    e1 = M @ (W @ y)

    def neg_ll(rho):
        e = e0 - rho * e1
        sigma2 = float(e @ e) / n
        A = I - rho * W
        sign, logdetA = np.linalg.slogdet(A)
        if sign <= 0 or sigma2 <= 0:
            return 1e12
        return 0.5 * n * np.log(2 * np.pi * sigma2) - logdetA + 0.5 * n

    res = optimize.minimize_scalar(neg_ll, bounds=(-0.99, 0.99), method="bounded",
                                   options={"xatol": 1e-5})
    rho = float(res.x)
    Wy = W @ y
    # OLS of (y - rho Wy) on X
    y_star = y - rho * Wy
    beta = np.linalg.solve(X.T @ X, X.T @ y_star)
    e = y_star - X @ beta
    sigma2 = float(e @ e) / max(n - p - 1, 1)
    cov_beta = sigma2 * XtX_inv
    se_beta = np.sqrt(np.maximum(np.diag(cov_beta), 0.0))

    return RichResult(payload={
        "estimate": beta.tolist(),
        "se": se_beta.tolist(),
        "rho": rho,
        "sigma2": sigma2,
        "n": int(n),
        "method": "SAR lag (ML, concentrated log-likelihood)",
    })


def cheatsheet():
    return "sarla: SAR lag model (ML)"


# CANONICAL TEST
# Same X, y, W as sarre canonical. Expect rho ~ small (data has no lag
# structure) and beta ~ [intercept, slope].

File: r-package/morie/R/sarla.R
# SPDX-License-Identifier: GPL-2.0-only
#' Spatial autoregressive lag model (SAR lag, ML).
#'
#' Y = rho W Y + X beta + eps,  eps ~ N(0, sigma2 I).
#' Concentrated log-likelihood in rho.
#'
#' @param x Design matrix (n by p).
#' @param y Response, length n.
#' @param w n-by-n weights matrix.
#' @return Named list: estimate, se, rho, sigma2, n, method.
#' @references Anselin (1988); Schabenberger & Gotway (2005), Ch 7.
#' @export
sarla <- function(x, y, w) {
  X <- as.matrix(x); y <- as.numeric(y); W <- as.matrix(w)
  n <- length(y); p <- ncol(X)
  if (nrow(X) != n || any(dim(W) != c(n, n)))
    stop("shape mismatch among x, y, w")
  I <- diag(n)
  XtX_inv <- solve(crossprod(X))
  M <- I - X %*% XtX_inv %*% t(X)
  e0 <- M %*% y
  e1 <- M %*% (W %*% y)
  neg_ll <- function(rho) {
    e <- e0 - rho * e1
    sigma2 <- as.numeric(sum(e ^ 2)) / n
    A <- I - rho * W
    det_sign <- determinant(A, logarithm = TRUE)
    if (det_sign$sign <= 0 || sigma2 <= 0) return(1e12)
    logdetA <- as.numeric(det_sign$modulus)
    0.5 * n * log(2 * pi * sigma2) - logdetA + 0.5 * n
  }
  res <- stats::optimize(neg_ll, interval = c(-0.99, 0.99))
  rho <- res$minimum
  Wy <- W %*% y
  y_star <- as.numeric(y - rho * Wy)
  beta <- as.numeric(XtX_inv %*% crossprod(X, y_star))
  e <- y_star - as.numeric(X %*% beta)
  sigma2 <- sum(e ^ 2) / max(n - p - 1, 1)
  cov_b <- sigma2 * XtX_inv
  se <- sqrt(pmax(diag(cov_b), 0))
  list(estimate = beta, se = se, rho = rho, sigma2 = sigma2, n = n,
       method = "SAR lag (ML, concentrated log-likelihood)")
}

# CANONICAL TEST  (with row-standardised path graph)

#' @rdname sarla
#' @keywords internal
#' @export
spatial_ar_lag <- sarla
